File: project/drivers/driver/python/src/scratchbird/telemetry.py
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

"""
OpenTelemetry-compatible Telemetry Module
Distributed tracing and metrics collection
"""

import time
import random
import re
from dataclasses import dataclass, field
from datetime import datetime
from typing import Dict, List, Optional, Callable
from threading import Lock
import uuid


@dataclass
class TelemetryConfig:
    """Configuration for telemetry"""
    enable_tracing: bool = True
    enable_metrics: bool = True
    enable_slow_query_log: bool = True
    slow_query_threshold_ms: int = 1000
    sanitize_queries: bool = True
    sample_rate: float = 1.0


@dataclass
class SpanContext:
    """Distributed tracing span context"""
    trace_id: str
    span_id: str
    parent_span_id: Optional[str]
    span_name: str
    start_time: float
    attributes: Dict[str, str] = field(default_factory=dict)
    
    @classmethod
    def new(cls, name: str) -> 'SpanContext':
        """Create a new root span"""
        return cls(
            trace_id=uuid.uuid4().hex,
            span_id=uuid.uuid4().hex[:16],
            parent_span_id=None,
            span_name=name,
            start_time=time.time()
        )
    
    @classmethod
    def child(cls, name: str, parent: 'SpanContext') -> 'SpanContext':
        """Create a child span"""
        return cls(
            trace_id=parent.trace_id,
            span_id=uuid.uuid4().hex[:16],
            parent_span_id=parent.span_id,
            span_name=name,
            start_time=time.time()
        )
    
    def with_attribute(self, key: str, value: str) -> 'SpanContext':
        """Add an attribute"""
        self.attributes[key] = value
        return self
    
    def elapsed_ms(self) -> int:
        """Get elapsed time in milliseconds"""
        return int((time.time() - self.start_time) * 1000)


@dataclass
class LatencyHistogram:
    """Latency distribution histogram"""
    ms_0_10: int = 0
    ms_10_100: int = 0
    ms_100_1000: int = 0
    ms_1000_10000: int = 0
    ms_over_10000: int = 0
    _lock: Lock = field(default_factory=Lock)
    
    def record(self, duration_ms: int) -> None:
        """Record a duration"""
        with self._lock:
            if duration_ms <= 10:
                self.ms_0_10 += 1
            elif duration_ms <= 100:
                self.ms_10_100 += 1
            elif duration_ms <= 1000:
                self.ms_100_1000 += 1
            elif duration_ms <= 10000:
                self.ms_1000_10000 += 1
            else:
                self.ms_over_10000 += 1
    
    def to_dict(self) -> Dict[str, int]:
        """Convert to dictionary"""
        with self._lock:
            return {
                'ms_0_10': self.ms_0_10,
                'ms_10_100': self.ms_10_100,
                'ms_100_1000': self.ms_100_1000,
                'ms_1000_10000': self.ms_1000_10000,
                'ms_over_10000': self.ms_over_10000
            }


@dataclass
class OperationMetrics:
    """Metrics for a specific operation"""
    count: int = 0
    total_time_ms: int = 0
    avg_time_ms: int = 0
    error_count: int = 0
    _lock: Lock = field(default_factory=Lock)
    
    def record(self, duration_ms: int, success: bool) -> None:
        """Record a metric"""
        with self._lock:
            self.count += 1
            self.total_time_ms += duration_ms
            self.avg_time_ms = self.total_time_ms // self.count
            if not success:
                self.error_count += 1
    
    def to_dict(self) -> Dict[str, int]:
        """Convert to dictionary"""
        with self._lock:
            return {
                'count': self.count,
                'total_time_ms': self.total_time_ms,
                'avg_time_ms': self.avg_time_ms,
                'error_count': self.error_count
            }


@dataclass
class SlowQueryLog:
    """Slow query log entry"""
    trace_id: str
    span_name: str
    duration_ms: int
    timestamp: datetime
    attributes: Dict[str, str]


@dataclass
class MetricsSnapshot:
    """Snapshot of current metrics"""
    total_queries: int
    successful_queries: int
    failed_queries: int
    total_query_time_ms: int
    latency_histogram: Dict[str, int]
    operation_metrics: Dict[str, Dict[str, int]]


class TelemetryCollector:
    """Collects spans and metrics"""
    
    def __init__(self, config: TelemetryConfig = None):
        self.config = config or TelemetryConfig()
        self._spans: List[SpanContext] = []
        self._spans_lock = Lock()
        self._total_queries = 0
        self._successful_queries = 0
        self._failed_queries = 0
        self._total_query_time_ms = 0
        self._metrics_lock = Lock()
        self._histogram = LatencyHistogram()
        self._operation_metrics: Dict[str, OperationMetrics] = {}
        self._op_metrics_lock = Lock()
        self._slow_queries: List[SlowQueryLog] = []
        self._slow_queries_lock = Lock()
    
    def start_span(self, name: str) -> Optional[SpanContext]:
        """Start a new span"""
        if not self.config.enable_tracing:
            return None
        
        # Sample rate check
        if random.random() > self.config.sample_rate:
            return None
        
        span = SpanContext.new(name)
        
        with self._spans_lock:
            self._spans.append(span)
            if len(self._spans) > 1000:
                self._spans.pop(0)
        
        return span
    
    def end_span(self, span: SpanContext, success: bool = True) -> None:
        """End a span and record metrics"""
        if not span or not self.config.enable_tracing:
            return
        
        duration_ms = span.elapsed_ms()
        self._record_query_metrics(span.span_name, duration_ms, success)
        
        if self.config.enable_slow_query_log and duration_ms > self.config.slow_query_threshold_ms:
            self._record_slow_query(span, duration_ms)
    
    def _record_query_metrics(self, operation: str, duration_ms: int, success: bool) -> None:
        """Record query metrics"""
        if not self.config.enable_metrics:
            return
        
        with self._metrics_lock:
            self._total_queries += 1
            if success:
                self._successful_queries += 1
            else:
                self._failed_queries += 1
            self._total_query_time_ms += duration_ms
        
        self._histogram.record(duration_ms)
        
        # Per-operation metrics
        with self._op_metrics_lock:
            if operation not in self._operation_metrics:
                self._operation_metrics[operation] = OperationMetrics()
            self._operation_metrics[operation].record(duration_ms, success)
    
    def _record_slow_query(self, span: SpanContext, duration_ms: int) -> None:
        """Record a slow query"""
        log = SlowQueryLog(
            trace_id=span.trace_id,
            span_name=span.span_name,
            duration_ms=duration_ms,
            timestamp=datetime.utcnow(),
            attributes=span.attributes.copy()
        )
        
        with self._slow_queries_lock:
            self._slow_queries.append(log)
            if len(self._slow_queries) > 100:
                self._slow_queries.pop(0)
    
    def get_metrics(self) -> MetricsSnapshot:
        """Get current metrics snapshot"""
        with self._metrics_lock:
            total = self._total_queries
            successful = self._successful_queries
            failed = self._failed_queries
            total_time = self._total_query_time_ms
        
        op_metrics = {}
        with self._op_metrics_lock:
            for op, metrics in self._operation_metrics.items():
                op_metrics[op] = metrics.to_dict()
        
        return MetricsSnapshot(
            total_queries=total,
            successful_queries=successful,
            failed_queries=failed,
            total_query_time_ms=total_time,
            latency_histogram=self._histogram.to_dict(),
            operation_metrics=op_metrics
        )
    
    def get_slow_queries(self) -> List[SlowQueryLog]:
        """Get slow query logs"""
        with self._slow_queries_lock:
            return self._slow_queries.copy()
    
    @staticmethod
    def sanitize_query(sql: str) -> str:
        """Sanitize query by removing sensitive data"""
        if not sql:
            return sql
        # Replace quoted strings
        return re.sub(r"'[^']*'", "'?'", sql)
    
    def export_prometheus_metrics(self) -> str:
        """Export metrics in Prometheus format"""
        m = self.get_metrics()
        h = m.latency_histogram
        
        lines = [
            "# HELP scratchbird_queries_total Total number of queries",
            "# TYPE scratchbird_queries_total counter",
            f"scratchbird_queries_total {m.total_queries}",
            "# HELP scratchbird_query_duration_ms Query duration histogram",
            "# TYPE scratchbird_query_duration_ms histogram",
            f"scratchbird_query_duration_ms_bucket{{le=\"10\"}} {h['ms_0_10']}",
            f"scratchbird_query_duration_ms_bucket{{le=\"100\"}} {h['ms_0_10'] + h['ms_10_100']}",
            f"scratchbird_query_duration_ms_bucket{{le=\"1000\"}} {h['ms_0_10'] + h['ms_10_100'] + h['ms_100_1000']}",
        ]
        
        return "\n".join(lines)


class TelemetrySpanGuard:
    """RAII guard for span lifecycle management"""
    
    def __init__(self, collector: TelemetryCollector, span: SpanContext):
        self.collector = collector
        self.span = span
        self.success = True
    
    def mark_failed(self) -> None:
        """Mark the span as failed"""
        self.success = False
    
    def __enter__(self) -> 'TelemetrySpanGuard':
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if exc_type is not None:
            self.success = False
        self.collector.end_span(self.span, self.success)

File: project/drivers/driver/ruby/lib/scratchbird/telemetry.rb
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

# ScratchBird Ruby Driver - OpenTelemetry Telemetry
# Copyright (c) 2025-2026 Dalton Calford

require 'securerandom'

module Scratchbird
  class TelemetryConfig
    attr_accessor :enable_tracing, :enable_metrics, :enable_slow_query_log,
                  :slow_query_threshold_ms, :sanitize_queries, :sample_rate
    
    def initialize(options = {})
      @enable_tracing = options.fetch(:enable_tracing, true)
      @enable_metrics = options.fetch(:enable_metrics, true)
      @enable_slow_query_log = options.fetch(:enable_slow_query_log, true)
      @slow_query_threshold_ms = options.fetch(:slow_query_threshold_ms, 1000)
      @sanitize_queries = options.fetch(:sanitize_queries, true)
      @sample_rate = options.fetch(:sample_rate, 1.0)
    end
  end
  
  class SpanContext
    attr_reader :trace_id, :span_id, :parent_span_id, :span_name, :start_time, :attributes
    
    def initialize(name, parent = nil)
      @trace_id = parent ? parent.trace_id : SecureRandom.hex(16)
      @span_id = SecureRandom.hex(8)
      @parent_span_id = parent ? parent.span_id : nil
      @span_name = name
      @start_time = Time.now
      @attributes = {}
    end
    
    def with_attribute(key, value)
      @attributes[key] = value
      self
    end
    
    def elapsed
      ((Time.now - @start_time) * 1000).to_i
    end
  end
  
  class LatencyHistogram
    attr_accessor :ms0_10, :ms10_100, :ms100_1000, :ms1000_10000, :ms_over_10000
    
    def initialize
      @ms0_10 = @ms10_100 = @ms100_1000 = @ms1000_10000 = @ms_over_10000 = 0
      @mutex = Mutex.new
    end
    
    def record(duration_ms)
      @mutex.synchronize do
        case duration_ms
        when 0..10 then @ms0_10 += 1
        when 11..100 then @ms10_100 += 1
        when 101..1000 then @ms100_1000 += 1
        when 1001..10000 then @ms1000_10000 += 1
        else @ms_over_10000 += 1
        end
      end
    end
  end
  
  class OperationMetrics
    attr_reader :count, :total_time_ms, :avg_time_ms, :error_count
    
    def initialize
      @count = @total_time_ms = @avg_time_ms = @error_count = 0
      @mutex = Mutex.new
    end
    
    def record(duration_ms, success)
      @mutex.synchronize do
        @count += 1
        @total_time_ms += duration_ms
        @avg_time_ms = @total_time_ms / @count
        @error_count += 1 unless success
      end
    end
  end
  
  class TelemetryCollector
    def initialize(config = TelemetryConfig.new)
      @config = config
      @spans = []
      @total_queries = @successful_queries = @failed_queries = @total_query_time_ms = 0
      @histogram = LatencyHistogram.new
      @operation_metrics = {}
      @slow_queries = []
      @mutex = Mutex.new
    end
    
    def start_span(name)
      return nil unless @config.enable_tracing
      return nil if rand > @config.sample_rate
      
      span = SpanContext.new(name)
      @mutex.synchronize do
        @spans << span
        @spans.shift if @spans.size > 1000
      end
      span
    end
    
    def end_span(span, success = true)
      return unless span && @config.enable_tracing
      duration_ms = span.elapsed
      record_query_metrics(span.span_name, duration_ms, success)
      if @config.enable_slow_query_log && duration_ms > @config.slow_query_threshold_ms
        record_slow_query(span, duration_ms)
      end
    end
    
    def metrics
      @mutex.synchronize do
        {
          total_queries: @total_queries,
          successful_queries: @successful_queries,
          failed_queries: @failed_queries,
          total_query_time_ms: @total_query_time_ms,
          latency_histogram: @histogram,
          operation_metrics: @operation_metrics.transform_values { |m| { count: m.count, avg_time_ms: m.avg_time_ms } }
        }
      end
    end
    
    def slow_queries
      @mutex.synchronize { @slow_queries.dup }
    end
    
    def self.sanitize_query(sql)
      return sql unless sql
      sql.gsub(/'[^']*'/, "'?'")
    end
    
    def export_prometheus_metrics
      m = metrics
      h = m[:latency_histogram]
      <<~METRICS
        # HELP scratchbird_queries_total Total number of queries
        # TYPE scratchbird_queries_total counter
        scratchbird_queries_total #{m[:total_queries]}
        # HELP scratchbird_query_duration_ms Query duration histogram
        # TYPE scratchbird_query_duration_ms histogram
        scratchbird_query_duration_ms_bucket{le="10"} #{h.ms0_10}
        scratchbird_query_duration_ms_bucket{le="100"} #{h.ms0_10 + h.ms10_100}
        scratchbird_query_duration_ms_bucket{le="1000"} #{h.ms0_10 + h.ms10_100 + h.ms100_1000}
      METRICS
    end
    
    private
    
    def record_query_metrics(operation, duration_ms, success)
      return unless @config.enable_metrics
      @mutex.synchronize do
        @total_queries += 1
        success ? @successful_queries += 1 : @failed_queries += 1
        @total_query_time_ms += duration_ms
        @histogram.record(duration_ms)
        @operation_metrics[operation] ||= OperationMetrics.new
        @operation_metrics[operation].record(duration_ms, success)
      end
    end
    
    def record_slow_query(span, duration_ms)
      @mutex.synchronize do
        @slow_queries << {
          trace_id: span.trace_id,
          span_name: span.span_name,
          duration_ms: duration_ms,
          timestamp: Time.now,
          attributes: span.attributes
        }
        @slow_queries.shift if @slow_queries.size > 100
      end
    end
  end
end
